geniushubclient/zone.py

```diff
@@ -39,9 +38,21 @@ class GeniusZone(GeniusBase):
     def __init__(self, zone_id, raw_json, hub) -> None:
         super().__init__(zone_id, raw_json, hub, ATTRS_ZONE)
 
+        self._properties = Properties()
```

Comment:
I strongly feel that the constructor should be for setting the class up, ready to be used. It shouldn't be processing data too.

Reasons for this: separation of concerns. Processing json can result in errors, so it is dangerous to put it in to the constructor. Additionally, once created, the Zone data should be getting updated on a regular basis, as the hub/server is pinged for updates. So you're going to need a publicly accessible update method anyway, I think it is then adding in an unnecessary extra pathway through the code.

Additionally, I don't think that anything else should be processing the json to retrieve data required for a method call, when this class is responsible for getting these values. You want all this parsing logic in one place, so if it ever needs changing, you've only got to do it one place. It also makes the classes up the hierarchy cleaner, as they won't have any json parsing code in them. They literally create a zone, then pass the json in to the update method.

Furthermore, by putting it in the constructor, you then require two constructors, as you've pointed out. That means that the zone class is going to have to check the hub version to decide which constructor to call. By passing the hub version in to the properties class update method, then it can decide which of its update methods to call. The zone doesn't have to care, as the properties class deals with that branching. (I will say that the api version could be in the constructor, if there is zero chance of the version changing during the lifetime of the class. That wasn't something I was sure of, so I've put it in the update method for now).

Whilst I think that this is good architectural design on its own, once you start writing unit tests, you really see the benefit. Look at the zone_get_v1_data_test file. 113 lines of code for one single test! And that's because the zone has the json in the constructor, and processes it. I only want to test that calling the data property on it results in a call to the Properties class to get its data. But I have to set up ~70 lines of JSON, because otherwise the class throws an error in the constructor. The same goes for the zone_update_test file. However, once there is zero json parsing in the data method (once all the classes have been written), there won't be a need to set up any json other than that directly required by the test file. Which in the case of the zone class, should be zero, as once it has no responsibility for parsing the json, you won't need any for its tests.

If I may turn the discussion on its head: what's the disadvantage in doing it this way? It seems that you're objecting to having to write 2 lines of code to set up and populate an object? Why is that an issue? Bear in mind, once all these classes are done, the GeniusZone class is going to be so much smaller, as all the business logic is going to be in the individual classes. The constructor will be setting up all the classes, the update method passing the json through to all the classes, and the data property asking the classes for the data.

I hope that all makes sense, and that I've made a compelling case, however, feel free to point out where you think I'm wrong.

Reply:
Aaaand.... was just thinking around this some more, particularly the updating an existing zone part; the client is going to need to get the zone id from the json to be able to pull the correct zone out of the dictionary, so, you're right about passing in the id in the constructor!

Reply:
Yep - IDs should invariably be in the constructor (the entity needs an ID).